specs/file_hierarchy_for_the_verification_of_os_artifacts.md

```diff
@@ -0,0 +1,561 @@
+
+#### Hierarchical trust
+
+(e.g. X.509, SSH+CA)
+
+Other systems have an inherent assumption of hierarchical delegation, formalized by cryptographic certifications.
+In such systems, globally or locally accepted _trust anchors_ are used as starting points in signature validation.
+Certificates and the delegation relations between them form a tree structure.
+Such hierarchical key structures are often referred to as [public key infrastructure] (PKI).
+
+This model requires explicitly defined _trust anchors_ in VOA.
+
+### Typical distribution format of verifers / short vs. long-lived keys
+
+VOA's goal is to provide a uniform representation of verifiers, to the degree that this is possible with differing technologies.
+To achieve this, verifiers are stored in a central file hierarchy that can be used with different technologies.
+
+Some concerns may be handled differently in supported technologies:
+
+- The separation between _trust anchors_ and _artifact verifiers_ may not be reasonable to model in separated directories, because the native representation of a given technology is to store both in a shared file, as a type of "certificate chain".
+- By default, VOA suggests that each verifier is stored in an individual file (that ideally reflects a unique identifier of that verifier in its filename).
+  However, in some technologies, separating verifiers into individual files may be uncommon or unreasonable.
+  In these technologies, bundles of verifiers may be stored in a shared file (this can apply equally to _artifact verifiers_ and _trust anchors_).
+- As an edge case, in some technologies, _artifact verifiers_ may be extremely short lived, so it may not be reasonable to store them in VOA.
+  In these cases, only a set of _trust anchors_ may be stored in VOA, while the _artifact verifiers_ are distributed with the signed artifacts.
+
+## File Hierarchy
+
+The _File Hierarchy for the Verification of OS Artifacts (VOA)_ is maintained as a directory structure on a system.
+It contains cryptographic public-key material for one or more technologies.
+
+The VOA hierarchy organizes _signature verifiers_ by [os], [purpose], [context] and [technology], and stores them in a directory structure of the form `$os/$purpose/$context/$technology/`.
+
+Vendors provide a set of _signature verifier_ files for their OS, in the VOA hierarchy format.
+
+Users of a VOA hierarchy (such as package installation software) can pick the relevant set of _signature verifiers_ for the verification of a specific artifact based on their location in the hierarchy.
+
+### Load paths
+
+VOA defines a list of _load paths_ for system mode and user mode, from which a hierarchy of _signature verifiers_ can be retrieved.
+
+Load paths when in system mode:
+
+- `/etc/voa/`
+- `/run/voa/`
+- `/usr/local/share/voa/`
+- `/usr/share/voa/`
+
+Load paths when in user mode:
+
+- `$XDG_RUNTIME_DIR/voa/`
+- `$XDG_DATA_HOME/voa/`
```

Comment:
should probably be `$XDG_DATA_DRS/voa/` no?

Reply:
Hmm, good point, but would we then not want to have XDG_DATA_DIRS in addition to XDG_DATA_HOME?

Looking at [the definition](https://specifications.freedesktop.org/basedir-spec/latest/#variables) I'm under the impression, that XDG_DATA_HOME has higher ordering than what is provided via XDG_DATA_DIRS?

Reply:
@poettering `XDG_DATA_DIRS` is an array of data directories similar to `$PATH`. So `XDG_DATA_HOME` is correct? You could include that `XDG_DATA_DIRS` should also be looked at.

Reply:
> If `$XDG_DATA_DIRS` is either not set or empty, a value equal to `/usr/local/share/:/usr/share/` should be used.

I think the specc needs to be clarified a bit more if you want to include `XDG_DATA_DIRS`.

Reff; https://specifications.freedesktop.org/basedir-spec/latest/

Reply:
see section 4 of the basedir spec. specs should reference $XDG_DATA_DIRS/foo/bar and that shall implicitly mean $XDG_DATA_HOME too, and if neither variable is set the default. i'd just use that. or in other words: no need to specify XDG_DATA_HOME among the search path list explicitly then.

Reply:
(i mean, you can make this explicit in your spec if you want, but it's kinda redundant to mention `$XDG_DATA_HOME` explicitly if you use the `$XDG_DATA_DIRS/foo/bar` wording, that's all I am saying. You could also add `as per section 4 of xdg spec` or so in your spec

Reply:
I've added support for XDG_CONFIG_DIRS and XDG_DATA_DIRS in https://github.com/uapi-group/specifications/pull/134/commits/20dcf7432b17d7bca04905eb3ddc8c64122e84fb